### ISSUE TITLE
'innerText' should not be nullable in most of Node's derived types

### ISSUE THREAD

**Issue by DanielRosenwasser** [microsoft/TypeScript#10315]
Compile the following with `--strictNullChecks` with the DOM available.

``` ts
const foo = document.createElement("div");
const s: string = foo.textContent;
```

Expected: No error
Actual: `Type 'string | null' is not assignable to type 'string'.`

From https://developer.mozilla.org/en-US/docs/Web/API/Node/textContent:

> - `textContent` returns `null` if the element is a [document](https://developer.mozilla.org/en-US/docs/DOM/document), a document type, or a notation. [...]
> - If the node is a CDATA section, a comment, a processing instruction, or a text node, textContent returns the text inside this node (the [`nodeValue`](https://developer.mozilla.org/en-US/docs/DOM/Node.nodeValue)).
> - For other node types, `textContent` returns the concatenation of the `textContent` attribute value of every child node, excluding comments and processing instruction nodes. This is an empty string if the node has no children.

Proposed fix: `textContent will be defined as nullable in`Node`, but overridden as`string` in most of its descendants.

**Comment by mhegazy** [comment]
PRs welcomed. You can find more information about contributing lib.d.ts fixes at https://github.com/Microsoft/TypeScript/blob/master/CONTRIBUTING.md#contributing-libdts-fixes.

**Comment by FranklinWhale** [comment]
Unfortunately `null` is always allowed when `textContent` is set:

https://dom.spec.whatwg.org/#dom-node-textcontent

I have tried to use getter and setter to fix this, but that seems not possible because "'get' and 'set' accessor must have the same type".

```ts
get textContent(): string;
set textContent(value: string | null);
````

**Comment by zanona** [comment]
Another test case:
```ts
const span : HTMElement | null = document.body.querySelector('span');
let value = '';
if (span instanceof HTMLElement) {
   value = span.textContent; // Type 'string | null' is not assignable to type 'string'. 
}
```
### Expected:
Valid, since anything subclass of `Element` type should always return a string.

**Comment by ExE-Boss** [comment]
@FranklinWhale That depends on <https://github.com/microsoft/TypeScript/issues/32821>.

**Comment by lionel-rowe** [comment]
Can't TypeScript just be a little stricter than the spec here, and disallow setting `textContent` to `null` on nodes where its value is always a string? I can't fathom why you'd ever even want to do that (as opposed to setting it to `""`, which has functionally identical behavior)

**Comment by Jamesernator** [comment]
> Can't TypeScript just be a little stricter than the spec here, and disallow setting `textContent` to `null` on nodes where its value is always a string?

It doesn't even need to, as far as I can tell the needed behaviour here [has worked since TS4.3](https://www.typescriptlang.org/play?ts=4.3.5#code/JYOwLgpgTgZghgYwgAgHIHsAmEBMyDeAUMicgOYRjKQAeYAwuuBOABQCUAXMgM5hSgyyAD7IQAVwA2kgNzFSPStQh1GzOqwBu3PgJBDRE6ezkBfQoVCRYiFAFFJEALYsweFZBCYeaLLgLyJBRUtAxMnmAcOvyCcqS8SqFqEVrRegZiUpImhOaE2AiScFAoCEx8yBDcDs6uOHKEZSAV2rwx+sgAvJUAdEnhrnJAA).

**Comment by Andarist** [comment]
This would be quite a nice improvement. The DOM types are heavily generated and it's not straightforward to tinker with them. Perhaps @saschanaz could advise how this could be achieved in practice there? I could give it a shot then.

**Comment by RyanCavanaugh** [comment]
Fixed in 5.9. AI said:
## Unable to Reproduce ❌

I attempted to reproduce this issue but could not confirm it still exists.

Test case used:
**test.ts:**
```typescript
const foo = document.createElement("div");
const s: string = foo.textContent;
```

**tsconfig.json:**
```json
{
    "compilerOptions": {
        "strictNullChecks": true,
        "lib": ["dom", "esnext"]
    }
}
```


**Command:** `tsc -p .`
**TypeScript Version:** Version 5.9.0-dev.20250722
**Exit Code:** 0

<details>
<summary>Compiler Output</summary>

**Stdout:**
```
(empty)
```

**Stderr:**
```
(empty)
```
</details>

**What you observed in the bug report:** An error occurs because 'textContent' was typed as 'string | null', making it not assignable to 'string'.
**What you expected:** No error occurs when assigning 'foo.textContent' to a 'string' variable, as 'textContent' should be 'string' in this context.
**What I observed when running tsc:** No error occurs when 'foo.textContent' is assigned to a 'string', as seen from the absence of compiler errors.

My reasoning:
> In the original bug report, the user expected no error when assigning `foo.textContent` to a `string` variable because they believed `textContent` should always be a `string` for certain node types. However, they observed an error because `textContent` was typed as `string | null`, which is not directly assignable to `string` without handling the `null` type.

In the current compiler output, there are no compiler errors when the given TypeScript code is compiled. This indicates that the assignment of `foo.textContent` to a `string` variable is not causing a type error. This suggests that the type of `textContent` has been adjusted or inferred in such a way that allows it to be directly assigned to a `string` without error in this context.

Therefore, the current behavior matches the expected behavior as originally described in the bug report. This indicates that the issue where `textContent` was incorrectly flagged as potentially `null` has been resolved for the context provided.

Thus, the bug has been fixed.


---

**Analysis Result:**
```json
{
  "bug_is_present": "no",
  "files": [
    {
      "filename": "test.ts",
      "content": "const foo = document.createElement(\"div\");\nconst s: string = foo.textContent;"
    },
    {
      "filename": "tsconfig.json",
      "content": "{\n    \"compilerOptions\": {\n        \"strictNullChecks\": true,\n        \"lib\": [\"dom\", \"esnext\"]\n    }\n}"
    }
  ],
  "user_reported_behavior": "An error occurs because 'textContent' was typed as 'string | null', making it not assignable to 'string'.",
  "expected_behavior": "No error occurs when assigning 'foo.textContent' to a 'string' variable, as 'textContent' should be 'string' in this context.",
  "current_observed_behavior": "No error occurs when 'foo.textContent' is assigned to a 'string', as seen from the absence of compiler errors.",
  "command_line": "tsc -p .",
  "tsc_version": "Version 5.9.0-dev.20250722",
  "compiler_output": {
    "stdout": "",
    "stderr": "",
    "exit_code": 0
  },
  "reasoning": "In the original bug report, the user expected no error when assigning `foo.textContent` to a `string` variable because they believed `textContent` should always be a `string` for certain node types. However, they observed an error because `textContent` was typed as `string | null`, which is not directly assignable to `string` without handling the `null` type.\n\nIn the current compiler output, there are no compiler errors when the given TypeScript code is compiled. This indicates that the assignment of `foo.textContent` to a `string` variable is not causing a type error. This suggests that the type of `textContent` has been adjusted or inferred in such a way that allows it to be directly assigned to a `string` without error in this context.\n\nTherefore, the current behavior matches the expected behavior as originally described in the bug report. This indicates that the issue where `textContent` was incorrectly flagged as potentially `null` has been resolved for the context provided.\n\nThus, the bug has been fixed."
}
```
